### PR TITLE
Adding a custom trigger guide

### PR DESCRIPTION
**Pull request by yuhgto**
Added a new app to show how to configure a custom trigger. The custom trigger is "Every 5 seconds, create an item."

### CODE REVIEW

File: apps/custom-trigger/src/app.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+const app = express();
+const port = process.env.PORT;
+// parse various different custom JSON types as JSON
+app.use(bodyParser.json())
+app.use(routes);
+app.listen(port, () => {
+    console.log(`Quickstart app listening at http://localhost:${port}`);
+    try {
+        triggerController.initialize(5);
```
Comment (vladmonday) [inline apps/custom-trigger/src/app.js:15]: Let's move this 5 to a constant

File: apps/custom-trigger/src/controllers/subscription-controller.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  console.log(`A new subscription was added. The webhook URL is: ${webhookUrl}`);
+
+  // Store subsciption URL and ID in the DB
+  // See /services/subscription-model-service.js to learn more about updating the databasse
+  try {
+    await subscriptionModelService.createSubscription(webhookUrl, subscriptionId);
+    return res.status(200).send({webhookId: subscriptionId});
+  }
+  catch(err) {
+    console.log(err);
```
Comment (vladmonday) [inline apps/custom-trigger/src/controllers/subscription-controller.js:17]: Need to respond here with status 500

File: apps/custom-trigger/src/controllers/subscription-controller.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+async function unsubscribe(req, res) {
+  // TODO: remove subscription from DB
+  const { webhookId } = req.body.payload;
+
+  try {
+    await subscriptionModelService.removeSubscription(webhookId);
+    return res.status(200).send({result: 'Unsubscribed successfully.'});
+  }
+  catch(err) {
+    console.log(err);
```
Comment (vladmonday) [inline apps/custom-trigger/src/controllers/subscription-controller.js:30]: Need to respond here with status 500

File: apps/custom-trigger/src/controllers/trigger-controller.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+        }
+    }, timeoutInSeconds * 1000);
+}
+
+async function invokeActiveTriggers() {
+    try {
+        const activeSubscriptions = await subscriptionModelService.getActiveSubscriptions();
+        let triggers = activeSubscriptions.map(async (subscription) => {
+            const webhookUrl = subscription.dataValues.webhook_url;
+            triggerService.callWebhookUrl(webhookUrl);
```
Comment (vladmonday) [inline apps/custom-trigger/src/controllers/trigger-controller.js:20]: I think we should have "return" it here

File: apps/custom-trigger/src/db/migrations/20200725221910-create-subscription.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+module.exports = {
+  up: (queryInterface, Sequelize) => {
+    return queryInterface.createTable('Subscriptions', {
+      id: {
+        allowNull: false,
+        autoIncrement: true,
+        primaryKey: true,
+        type: Sequelize.INTEGER
+      },
+      id: {
```
Comment (vladmonday) [inline apps/custom-trigger/src/db/migrations/20200725221910-create-subscription.js:11]: Why do we have here another id?

File: apps/custom-trigger/src/db/migrations/20200725221910-create-subscription.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+        type: Sequelize.STRING
+      },
+      active: {
+        type: Sequelize.BOOLEAN
+      },
+      created_at: {
```
Comment (vladmonday) [inline apps/custom-trigger/src/db/migrations/20200725221910-create-subscription.js:17]: why we need to have both created_at and createdAt?

File: apps/custom-trigger/src/db/models/index.js [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+const db = {};
+
+let sequelize;
+if (config.use_env_variable) {
+  sequelize = new Sequelize(process.env[config.use_env_variable], config);
+} else {
+  sequelize = new Sequelize(config.database, config.username, config.password, config);
+}
+
+fs
```
Comment (vladmonday) [inline apps/custom-trigger/src/db/models/index.js:18]: Why do we read the config file with fs and not just require it? node can require JSON files natively

File: apps/custom-trigger/src/middlewares/authentication.js [hunk 8]
```diff
@@ -0,0 +1,23 @@
+const jwt = require('jsonwebtoken');
+const SHARED_SECRET = process.env.MONDAY_SIGNING_SECRET;
+
+async function authenticationMiddleware(req, res, next) {
+  try {
+    let { authorization } = req.headers;
```
Comment (vladmonday) [inline apps/custom-trigger/src/middlewares/authentication.js:6]: We really need to move this middleware to the SDK

File: apps/custom-trigger/src/routes/subscription.js [hunk 9]
```diff
@@ -0,0 +1,12 @@
+// This file defines which functions are invoked when a particular route/endpoint is called.
+
+const express = require('express');
+const router = express.Router();
+
+const subscriptionController = require('../controllers/subscription-controller.js');
+const authenticationMiddleware = require('../middlewares/authentication').authenticationMiddleware;
+
+router.post('/subscriptions/subscribe', subscriptionController.subscribe);
```
Comment (vladmonday) [inline apps/custom-trigger/src/routes/subscription.js:9]: We need to add authneticationMiddleware to routes, like here: https://github.com/mondaycom/welcome-apps/blob/master/apps/quickstart-integrations/src/routes/transformation.js#L7

File: apps/custom-trigger/src/services/monday-service.js [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+        items (ids: $itemId) {
+          column_values(ids:$columnId) {
+            value
+          }
+        }
+      }`;
+      const variables = { columnId, itemId };
+
+      const response = await mondayClient.api(query, { variables });
+      return response.data.items[0].column_values[0].value;
```
Comment (vladmonday) [inline apps/custom-trigger/src/services/monday-service.js:19]: dangerous, need to check items.length and column_values.length, to not fail on the empty arrays

File: apps/custom-trigger/src/services/monday-service.js [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+    } catch (err) {
+      console.log(err);
```
Comment (vladmonday) [inline apps/custom-trigger/src/services/monday-service.js:21]: need to throw error higher
throw err;

File: apps/custom-trigger/src/services/monday-service.js [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+          id
+        }
+      }
+      `;
+      const variables = { boardId, columnId, itemId, value };
+
+      const response = await mondayClient.api(query, { variables });
+      return response;
+    } catch (err) {
+      console.log(err);
```
Comment (vladmonday) [inline apps/custom-trigger/src/services/monday-service.js:40]: need to throw error higher
throw err;

File: apps/custom-trigger/src/db/models/subscription.js [hunk 13]
```diff
@@ -0,0 +1,18 @@
+'use strict';
+module.exports = (sequelize, DataTypes) => {
+  const Subscription = sequelize.define('Subscription', {
+    id: {
+      type: DataTypes.STRING,
+      primaryKey: true
+    },
+    active: DataTypes.BOOLEAN,
+    created_at: DataTypes.DATE,
```
Comment (vladmonday) [inline apps/custom-trigger/src/db/models/subscription.js:9]: let's use Sequelize default timestamps:
      createdAt: Sequelize.DATE,
      updatedAt: Sequelize.DATE,

https://sequelize.org/v5/manual/models-definition.html (TIMESTAMPS section)

**Comment by yuhgto** [comment]
Should be good to go! Ignore the "Added oauth example repo" commit, i added a new app and then removed it to keep this branch clean.

**Comment by yuhgto** [comment]
@edogr1 -- can you help me review this PR? It's a basic custom trigger recipe guide.

**Comment by yuhgto** [comment]
Closing as this PR is stale (lots of conflicts).

**Pull request closed without merging**